### PR TITLE
[OptimizeForJS] Optimize 64-bit div by constant

### LINKED ISSUE #4054: Magic division by constant for OptimizeForJS. Does it make sense?
I'm working on a new rule for OptimizeForJS that optimizes the division / modulo by constant for 64-bit types. Since for this particular case it is too expensive to call and execute a helper (polyfill) function for full division. According to my simple tests this gives up to 7x performance gain if the high part of dividend is not zero. The PR will come out pretty huge, so I'm wondering if it's worth it.

Rule will rewrite this `i64(x) / C` to something like this in simplified form `mulh64(x, mag.M) >> mag.S`

### PR DESCRIPTION
**Pull request by MaxGraey**
Optimize only for division. (Opts for reminder will be in separate PR later).

JavaScript does not have 64-bit integer types (except BigInt which is not portable and slow), so most operations such as division must be replaced by emulated routine of the corresponding operation. In this case it is `__wasm_i64_udiv` and `__wasm_i64_sdiv` which do quite a lot of work and contain loops. They are called even for such cases when the divisor is a constant (including zero). We can avoid this and use well-known algorithms in v8 or LLVM / GCC for that and replace expensive calls with cheaper multiplication and shift and add also some fast paths for special cases, e.g. for case when sign divisor is power of two value.

I evaluated the acceleration that this transformation gives and for a divisible whose high part of word is not zero. This speedup unsigned division up to 7 times.

For unsigned:
```rust
u64(x) / 3   
```
=>   
```rust
if (high_part(x) == 0) {
  return u64(u32(x) / 3) // fast path
} else {
  return mulh(x, -6148914691236517205) >> 1
}
```
before was: `__wasm_i64_udiv(x, 3)`

For signed (power of two):
```rust
i64(x) /  4    =>     ((x < 0 ? x + 3 : x) >> 2)
```
```rust
i64(x) / -8    =>    -((x < 0 ? x + 7 : x) >> 3)
```

For signed (non-power of two):

```rust
i64(x) /  3
```
=>
```rust
if (high_part(x) == 0) {
  return i64(i32(x) / 3) // fast path
} else {
  return mulh(x, 6148914691236517206) + (u64(x) >> 63)
}
```

Fix #4054

### CODE REVIEW

**Comment by MaxGraey** [comment]
@kripken fuzz_opt.py fuzzer (seed 17782795700923856759) failed on this reduced example:
```wat
(func (export "test") (param $0 i64) (result i64)
  (local $2 i64)
  (i64.div_u
   (local.get $2) ;; basically it's 0
   (i64.const 4294967294)
  )
 )
```
which should produce zero result in any case.

Current optimization rewrite this to:
```wat 
 (func (export "test_opt") (param $x i64) (result i64)
  (local $y i64)
  (local $2 i64)
  (if (result i64)
  (i64.eqz
   (i64.shr_u
    (i64.shr_u
     (local.get $y)
     (i64.const 1)
    )
    (i64.const 32)
   )
  )
  (i64.extend_i32_u
   (i32.div_u
    (i32.wrap_i64
     (i64.shr_u
      (local.get $y)
      (i64.const 1)
     )
    )
    (i32.const 2147483647)
   )
  )
  (i64.shr_u
   (local.tee $2
    (call $__wasm_i64_mulh
     (i64.shr_u
      (local.get $y)
      (i64.const 1)
     )
     (i64.const -9223372032559808509)
    )
   )
   (i64.const 30)
  )
 )
 )
 
 (func $__wasm_i64_mulh (type $0) (param $var$0 i64) (param $var$1 i64) (result i64)
  (local $var$2 i64)
  (local $var$3 i64)
  (local.set $var$0
   (i64.add
    (i64.mul
     (local.tee $var$2 (i64.and (local.get $var$1) (i64.const 4294967295)))
     (local.tee $var$3 (i64.shr_u (local.get $var$0) (i64.const 32)))
    )
    (i64.shr_u
     (i64.mul
      (local.get $var$2)
      (local.tee $var$2 (i64.and (local.get $var$0) (i64.const 4294967295)))
     )
     (i64.const 32)
    )
   )
  )
  (i64.add
   (i64.add
    (i64.mul
     (local.get $var$3)
     (local.tee $var$1 (i64.shr_u (local.get $var$1) (i64.const 32)))
    )
    (i64.shr_u (local.get $var$0) (i64.const 32))
   )
   (i64.shr_u
    (i64.add
     (i64.mul (local.get $var$1) (local.get $var$2))
     (i64.and (local.get $var$0) (i64.const 4294967295))
    )
    (i64.const 32)
   )
  )
 )
```

which also produce zero for any case.

So I'm a little confused. Could this be a fuzzer problem or am I missing something?

**Comment by kripken** [comment]
I am not sure I understand this. It uses code from V8 to optimize constant division at compile time - but wouldn't V8 and other engines do it at run time anyhow? Is there a benefit to doing it in the tools vs the VM?

**Comment by MaxGraey** [comment]
Yes, v8 uses this. The problem is that 64-bit division in JS turns into a rather expensive call which can be avoided by applying the same practices used in engines and LLVM (replacing division with multiplication and shift). I made benchmarks and found out that this speeds up division by a constant for 64-bit values up to 7 times for ASM.js code.

**Comment by MaxGraey** [comment]
Fuzzed until `ITERATION: 57567`

**Comment by MaxGraey** [comment]
I've updated the description. I hope the motivation is clearer now

**Comment by kripken** [comment]
Oh, thanks, I missed that this was for JS specifically. Yes, that seems like it could be useful.

File: src/passes/OptimizeForJS.cpp [hunk 1]
```diff
@@ -17,29 +17,94 @@
...
+
+    if (requireMulhIntrinsic) {
+      Module intrinsics;
+      std::string input(IntrinsicsModuleWast);
+      SExpressionParser parser(const_cast<char*>(input.c_str()));
+      Element& root = *parser.root;
+      SExpressionWasmBuilder builder(intrinsics, *root[0], IRProfile::Normal);
+      auto* func = intrinsics.getFunction(WASM_I64_MUL_HIGH);
+      doWalkFunction(ModuleUtils::copyFunction(func, *module));
+      requireMulhIntrinsic = false;
```
Comment (kripken) [inline src/passes/OptimizeForJS.cpp:54]: why set this to false?

File: src/passes/OptimizeForJS.cpp [hunk 2]
```diff
@@ -17,29 +17,94 @@
...
+    // {
+    //   // i64(x) % C   ==>   mulh(x, M') >> S'
+    //   //   where M' and S' are magic constants
+    //   Const* c;
+    //   Expression* x;
+    //   if (matches(curr, binary(RemS, any(&x), i64(&c)))) {
+    //     // requireMulhIntrinsic = true;
+    //     rewriteRemByConstS64(x, c->value.geti64());
+    //   }
+    // }
```
Comment (kripken) [inline src/passes/OptimizeForJS.cpp:107]: are the comments TODOs?

File: src/passes/wasm-intrinsics.wat [hunk 3]
```diff
@@ -142,6 +143,41 @@
    (local.get $var$1)
   )
  )
+ (func $__wasm_i64_mulh (; 5 ;) (type $0) (param $var$0 i64) (param $var$1 i64) (result i64)
```
Comment (kripken) [inline src/passes/wasm-intrinsics.wat:146]: was this generated using Rust like the comment on lines 1-6 says? Or some other way?

File: src/support/div-by-const.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
+// Copyright 2014 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
```
Comment (kripken) [inline src/support/div-by-const.cpp:3]: this LICENSE file should be included. if it's very short, maybe it can just be embedded here. otherwise, it should be a file on the side, which means probably there should be a new subdir for this, so it does not appear to apply to any other files than it should.

**Review by kripken — COMMENTED** [review]
Overall this does add a lot of complexity. But the speedup is significant as you said.

Do you have an idea of whether this happens in real-world code? Or is the worry theoretical for now?

**Comment by kripken** [comment]
To be clear, I think I can be convinced this is worth the complexity, but I'm not sure yet.

File: src/support/div-by-const.cpp (hunk 4 above)
Comment (MaxGraey) [reply src/support/div-by-const.cpp]: Sure, make sense. I just wanted to clarify this nuance.
Done!

File: src/passes/OptimizeForJS.cpp (hunk 1 above)
Comment (MaxGraey) [reply src/passes/OptimizeForJS.cpp:54]: yeah. This unnecessary.

File: src/passes/wasm-intrinsics.wat (hunk 3 above)
Comment (MaxGraey) [reply src/passes/wasm-intrinsics.wat:146]: It's generated from AssemblyScript file:
```ts
function umulh64(u: u64, v: u64): u64 {
  var u0 = u & 0xFFFFFFFF;
  var v0 = v & 0xFFFFFFFF;

  var u1 = u >> 32;
  var v1 = v >> 32;

  var l = u0 * v0;
  var t = u1 * v0 + (l >> 32);
  var w = u0 * v1 + (t & 0xFFFFFFFF);

  t >>= 32;
  w >>= 32;

  return u1 * v1 + t + w;
}
```
Rust generate more ore less the same codegen. But I could fork original gist and add Rust's solution

**Comment by MaxGraey** [comment]
> Do you have an idea of whether this happens in real-world code? Or is the worry theoretical for now?

Dividing by a constant is a fairly common operation. Perhaps even more frequent than the more common operation when the divisor is a more complex expression. The question, of course, is how often in real code we work with 64-bit values and use asmjs target for old browsers for example. I think it depends a lot on the application. Usually it is cryptography or compression/decompression algorithms. Also division is always involved in de-serializations (conv a number into a string).

File: src/passes/OptimizeForJS.cpp (hunk 2 above)
Comment (MaxGraey) [reply src/passes/OptimizeForJS.cpp:107]: Yes. Added TODO comment

File: src/passes/wasm-intrinsics.wat (hunk 3 above)
Comment (kripken) [reply src/passes/wasm-intrinsics.wat:146]: I think it's ok as it is. Just please update the lines of documentation there, and link to this code sample there (can link to this PR).

**Comment by kripken** [comment]
The reasoning makes sense. I don't like the complexity, but given the large speed I don't object.

If there was something in the middle - much simpler, but not as fast - I'd prefer that, but I can't think of anything.

File: test/wasm2js/i64-add-sub.2asm.js.opt [hunk 5]
```diff
@@ -21,7 +21,7 @@ function asmFunc(env) {
  }
  
  function legalstub$2($0, $1, $2, $3, $4, $5) {
-  return ($4 | 0) == ($0 - $2 | 0) & ($5 | 0) == ($1 - (($0 >>> 0 < $2 >>> 0) + $3 | 0) | 0);
+  return ($0 - $2 | 0) == ($4 | 0) & ($1 - (($0 >>> 0 < $2 >>> 0) + $3 | 0) | 0) == ($5 | 0);
  }
```
Comment (kripken) [inline test/wasm2js/i64-add-sub.2asm.js.opt:25]: do you know why this changes?

File: test/lit/passes/optimize-for-js.wast [hunk 6]
```diff
@@ -80,4 +84,693 @@
...
+ ;; Unsigned div by const
+
+ ;; CHECK:      (func $div-unsigned-by0-i64 (param $x i64) (result i64)
+ ;; CHECK-NEXT:  (i64.const 0)
+ ;; CHECK-NEXT: )
+ (func $div-unsigned-by0-i64 (param $x i64) (result i64)
+  (i64.div_u
+   (local.get $x)
+   (i64.const 0)
+  )
```
Comment (kripken) [inline test/lit/passes/optimize-for-js.wast:97]: This doesn't make sense to me. Division by zero should trap, so why is it turned into 0?

File: test/wasm2js/i64-add-sub.2asm.js.opt (hunk 5 above)
Comment (MaxGraey) [reply test/wasm2js/i64-add-sub.2asm.js.opt:25]: I also added extra optimization pass after optimize-for-js:
https://github.com/WebAssembly/binaryen/pull/4055/files#diff-503ab0ba884eaf72c574346fa45a6875959b6b1b3c90787a7d3ce4d7f6b464f6R348

this necessary due to this rule skip some cases which already can handled by "optimize-instructions" pass

File: src/passes/wasm-intrinsics.wat (hunk 3 above)
Comment (MaxGraey) [reply src/passes/wasm-intrinsics.wat:168]: Done!

**Comment by kripken** [comment]
In another PR the discussion suggested that JS opts are not a priority and we should focus on wasm ones, and on getting AS to use the wasm build. I wanted to check if we are in agreement in that?

If so then I think this PR is an example of something that I think may not be worth focusing on, given the complexity, and the low benefit if we switch AS to use wasm anyhow.

**Comment by MaxGraey** [comment]
Sure. But how about more simpler PRs? Like #4078 and #4083?
